lib/std/math/scalbn.zig

```diff
@@ -9,89 +9,68 @@
 // https://git.musl-libc.org/cgit/musl/tree/src/math/scalbnf.c
 // https://git.musl-libc.org/cgit/musl/tree/src/math/scalbn.c
 
-const std = @import("../std.zig");
+const std = @import("std");
 const math = std.math;
+const assert = std.debug.assert;
 const expect = std.testing.expect;
 
 /// Returns x * 2^n.
 pub fn scalbn(x: anytype, n: i32) @TypeOf(x) {
-    const T = @TypeOf(x);
-    return switch (T) {
-        f32 => scalbn32(x, n),
-        f64 => scalbn64(x, n),
-        else => @compileError("scalbn not implemented for " ++ @typeName(T)),
-    };
-}
-
-fn scalbn32(x: f32, n_: i32) f32 {
-    var y = x;
-    var n = n_;
+    var x_ = x;
+    var n_ = n;
 
-    if (n > 127) {
-        y *= 0x1.0p127;
-        n -= 127;
-        if (n > 1023) {
-            y *= 0x1.0p127;
-            n -= 127;
-            if (n > 127) {
-                n = 127;
-            }
-        }
-    } else if (n < -126) {
-        y *= 0x1.0p-126 * 0x1.0p24;
-        n += 126 - 24;
-        if (n < -126) {
-            y *= 0x1.0p-126 * 0x1.0p24;
-            n += 126 - 24;
-            if (n < -126) {
-                n = -126;
-            }
-        }
+    const T = @TypeOf(x_);
+    if (@typeInfo(T) != .Float) {
+        @compileError("scalbn not implemented for " ++ @typeName(T));
     }
 
-    const u = @intCast(u32, n +% 0x7F) << 23;
-    return y * @bitCast(f32, u);
-}
-
-fn scalbn64(x: f64, n_: i32) f64 {
-    var y = x;
-    var n = n_;
+    // calculate float limits
+    const TBits = std.meta.Int(.unsigned, @bitSizeOf(T));
+    const mantissa_size = @ctz(TBits, @bitCast(TBits, @as(T, 1.0)));
+    const exponent_max = @intCast(@TypeOf(n_), @bitCast(TBits, @as(T, 1.0)) >> mantissa_size);
+    const exponent_min = (-exponent_max) + 1; // +1 as lowest is reserved for subnormals
 
-    if (n > 1023) {
-        y *= 0x1.0p1023;
-        n -= 1023;
-        if (n > 1023) {
-            y *= 0x1.0p1023;
-            n -= 1023;
-            if (n > 1023) {
-                n = 1023;
-            }
+    // scale `n_` within floating point limits, if possible
+    // second pass is possible due to subnormal range
+    // third pass always results in +/-0.0 or +/-inf
+    const scale_max = @bitCast(T, @intCast(TBits, exponent_max * 2) << mantissa_size);
+    const scale_min = @bitCast(T, @as(TBits, 0b1) << mantissa_size);
+    if (n_ > exponent_max) {
+        x_ *= scale_max;
+        n_ -= exponent_max;
+        if (n_ > exponent_max) {
+            x_ *= scale_max;
+            n_ -= exponent_max;
+            if (n_ > exponent_max) n_ = exponent_max;
         }
-    } else if (n < -1022) {
-        y *= 0x1.0p-1022 * 0x1.0p53;
-        n += 1022 - 53;
-        if (n < -1022) {
-            y *= 0x1.0p-1022 * 0x1.0p53;
-            n += 1022 - 53;
-            if (n < -1022) {
-                n = -1022;
-            }
+    } else if (n_ < exponent_min) {
+        x_ *= scale_min;
```

Comment:
The rounding problem is still there, check out [this](https://git.musl-libc.org/cgit/musl/commit/src/math/scalbn.c?id=8c44a060243f04283ca68dad199aab90336141db) musl commit.